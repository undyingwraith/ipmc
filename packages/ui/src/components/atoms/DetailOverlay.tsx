import { Backdrop } from '@mui/material';
import { ComponentChildren } from 'preact';

export function DetailOverlay(props: {
	children: ComponentChildren;
}) {
	return (
		<Backdrop
			open={true}
			sx={{
				marginTop: '64px',
				width: '100vw',
				height: 'calc(100vh - 64px)',
			}}
		>
			{props.children}
		</Backdrop>
	);
}
