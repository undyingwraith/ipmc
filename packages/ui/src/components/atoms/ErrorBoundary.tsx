import { Alert } from '@mui/material';
import React, { PropsWithChildren } from 'react';

export class ErrorBoundary extends React.Component<PropsWithChildren<{}>, { hasError: boolean; }> {
	constructor(props: PropsWithChildren<{}>) {
		super(props);
		this.state = { hasError: false };
	}

	static getDerivedStateFromError(error: any) {
		// Update state so the next render will show the fallback UI.
		return { hasError: true };
	}

	componentDidCatch(error: Error, info: any) {
		console.error(error, info);
	}

	render() {
		if (this.state.hasError) {
			return (
				<Alert severity="error">Error</Alert>
			);
		}

		return this.props.children;
	}
}
