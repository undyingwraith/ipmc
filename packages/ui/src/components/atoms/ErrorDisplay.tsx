import { Alert, Typography } from '@mui/material';
import React from 'react';

export function ErrorDisplay(props: { error: Error; }) {
	const { error } = props;

	return (
		<Alert severity={'error'}>
			<Typography data-testid={'error-message'}>{error.name}: {error.message}</Typography>
		</Alert>
	);
}
