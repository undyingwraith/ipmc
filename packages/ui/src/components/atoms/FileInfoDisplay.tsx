import { Box, Stack, Typography } from '@mui/material';
import { IFileInfo, isPosterFeature } from 'ipmc-interfaces';
import React from 'react';
import { useFileUrl } from '../../hooks';
import { useTitle } from '../../hooks/useTitle';

export function FileInfoDisplay(props: { file: IFileInfo; }) {
	const { file } = props;
	const name = useTitle(file);
	const title = (
		<Typography variant={'h3'}>{name}</Typography>
	);
	const posterUrl = useFileUrl(isPosterFeature(file) && file.posters.length > 0 ? file.posters[0]?.cid : undefined);

	return isPosterFeature(file) && file.posters.length > 0 ? (
		<Stack direction={'row'} spacing={1}>
			<img src={posterUrl.value} style={{ height: 250, flexGrow: 0 }} />
			{title}
		</Stack>
	) : (
		<Box>
			{title}
		</Box>
	);
}
