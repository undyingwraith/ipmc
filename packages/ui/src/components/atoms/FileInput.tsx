import { Box, Button } from '@mui/material';
import { ReadonlySignal, Signal, useComputed } from '@preact/signals';
import { useTranslation } from '../../hooks';

interface IFileInputProps {
	label?: ReadonlySignal<string>;
	value: Signal<File[]>;
	accept?: string;
}

export function FileInput(props: IFileInputProps) {
	const _t = useTranslation();

	return (
		<Box>
			<Button
				variant={'contained'}
				component={'label'}
			>
				{props.label?.value ?? _t('UploadFile')}
				<input
					type={'file'}
					hidden
					accept={props.accept}
					onChange={(ev) => {
						props.value.value = Array.from(ev.currentTarget.files!);
					}}
				/>
			</Button>
			{useComputed(() => props.value.value.map((f) => (
				<Box>
					{f.name}
				</Box>
			)))}
		</Box>
	);
}
