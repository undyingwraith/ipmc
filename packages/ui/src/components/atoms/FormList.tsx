import { Button, Typography } from '@mui/material';
import Grid from '@mui/material/Grid2';
import { ReadonlySignal, Signal, useComputed } from '@preact/signals-react';
import { useTranslation } from '@src/hooks';
import React, { ReactNode } from 'react';

interface IFormListProps<TData> {
	renderControl: (data: Signal<TData>) => ReactNode;
	values: Signal<Signal<TData>[]>;
	createItem: () => TData;
	label: ReadonlySignal<string>;
}

export function FormList<TData>(props: IFormListProps<TData>) {
	const _t = useTranslation();

	return (
		<Grid container spacing={1}>
			<Grid size={8}>
				<Typography>{props.label}</Typography>
			</Grid>
			<Grid size={4}>
				<Button
					fullWidth={true}
					onClick={() => {
						props.values.value = [...props.values.value, new Signal(props.createItem())];
					}}
				>
					{_t('Add')}
				</Button>
			</Grid>
			{useComputed(() => props.values.value.map((item, index) => (
				<Grid size={12}>
					{props.renderControl(item)}
					<Button onClick={() => {
						props.values.value = props.values.value.toSpliced(index, 1);
					}}>{_t('Remove')}</Button>
				</Grid>
			)))}
		</Grid>
	);
}
