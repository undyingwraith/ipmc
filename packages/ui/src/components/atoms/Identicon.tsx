import React from 'react';
import { minidenticon } from 'minidenticons';

export function Identicon(props: { value: string; }) {
	return <img src={'data:image/svg+xml;utf8,' + encodeURIComponent(minidenticon(props.value))} width={25} height={25} />;
}
