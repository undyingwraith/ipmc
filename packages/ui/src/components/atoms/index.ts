export { ErrorBoundary } from './ErrorBoundary';
export { FileInput } from './FileInput';
export { FormList } from './FormList';
export { Identicon } from './Identicon';
export { ImageView } from './ImageView';
export { Loader } from './Loader';
export { SelectInput } from './SelectInput';
export { Spacer } from './Spacer';
export { TextInput } from './TextInput';
export { ThemeToggle } from './ThemeToggle';
