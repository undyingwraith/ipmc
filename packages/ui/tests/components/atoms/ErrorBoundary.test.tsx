import { render } from '@testing-library/preact';
import { ErrorBoundary } from 'src/components/atoms';
import { describe, expect, test } from 'vitest';

describe('ErrorBoundary', () => {
	test('Renders error message on error', () => {
		function TestComponent() {
			throw new Error('Sample error');
			return <div />;
		}

		const t = render(
			<ErrorBoundary>
				<TestComponent />
			</ErrorBoundary>
		);

		console.log('render');

		expect(t).toMatchSnapshot();
	});

	test('Renders component if no error is present', () => {
		const t = render(
			<ErrorBoundary>
				This is a test
			</ErrorBoundary>
		);

		expect(t.container).toMatchSnapshot();
	});
});
