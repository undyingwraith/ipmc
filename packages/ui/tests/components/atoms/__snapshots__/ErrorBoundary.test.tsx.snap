// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorBoundary > Renders component if no error is present 1`] = `
<div>
  This is a test
</div>
`;
